packages/multi-test/src/staking.rs

```diff
@@ -15,12 +29,240 @@ pub enum StakingSudo {
+    type ExecT = StakingMsg;
+    type QueryT = StakingQuery;
+    type SudoT = StakingSudo;
 
-pub type FailingDistribution = FailingModule<DistributionMsg, Empty, Empty>;
+    fn execute<ExecC, QueryC>(
+        &self,
+        _api: &dyn Api,
+        storage: &mut dyn Storage,
+        _router: &dyn CosmosRouter<ExecC = ExecC, QueryC = QueryC>,
+        _block: &BlockInfo,
+        sender: Addr,
+        msg: StakingMsg,
+    ) -> AnyResult<AppResponse> {
+        let mut staking_storage = prefixed(storage, NAMESPACE_STAKING);
+        match msg {
+            StakingMsg::Delegate { validator, amount } => {
+                let events = vec![Event::new("delegate")
+                    .add_attribute("recipient", &validator)
+                    .add_attribute("sender", &sender)
+                    .add_attribute("amount", format!("{}{}", amount.amount, amount.denom))];
+                self.add_stake(&mut staking_storage, sender, vec![amount])?;
+                Ok(AppResponse { events, data: None })
+            }
+            StakingMsg::Undelegate { validator, amount } => {
+                let events = vec![Event::new("undelegate")
+                    .add_attribute("from", &validator)
+                    .add_attribute("to", &sender)
+                    .add_attribute("amount", format!("{}{}", amount.amount, amount.denom))];
+                self.remove_stake(&mut staking_storage, sender, vec![amount])?;
+                Ok(AppResponse { events, data: None })
+            }
+            m => bail!("Unsupported staking message: {:?}", m),
+        }
+    }
 
-impl Distribution for FailingDistribution {}
+    fn sudo<ExecC, QueryC>(
+        &self,
+        _api: &dyn Api,
+        _storage: &mut dyn Storage,
+        _router: &dyn CosmosRouter<ExecC = ExecC, QueryC = QueryC>,
+        _block: &BlockInfo,
+        msg: StakingSudo,
+    ) -> AnyResult<AppResponse> {
+        bail!("Unsupported staking sudo message: {:?}", msg)
+    }
+
+    fn query(
+        &self,
+        api: &dyn Api,
+        storage: &dyn Storage,
+        _querier: &dyn Querier,
+        _block: &BlockInfo,
+        request: StakingQuery,
+    ) -> AnyResult<Binary> {
+        let staking_storage = prefixed_read(storage, NAMESPACE_STAKING);
+        match request {
+            StakingQuery::Delegation {
+                delegator,
+                validator,
+            } => {
+                // for now validator is ignored, as I want to support only one validator
+                let delegator = api.addr_validate(&delegator)?;
+                let stakes = match self.get_stakes(&staking_storage, &delegator) {
+                    Ok(stakes) => stakes[0].clone(),
+                    Err(_) => {
+                        let response = DelegationResponse { delegation: None };
+                        return Ok(to_binary(&response)?);
+                    }
+                };
+                // set fixed reward ratio 1:10 per delegated amoutn
```

Comment:
This should definitely be configurable somehow (via sudo or via custom "setup" calls).
Basically, nothing is free, you set it up defining how much rewards, then contract withdraws.
Can simulate contract withdrawing twice in one block, with 1 day between, with.1 year between that way